nixos/podman-dnsname: init

roberth opened the pull request:
###### Motivation for this change

Allow containers to contact each other by name.
Name resolution in container networks is rather useful.

This makes podman a near-feature complete alternative to docker, unblocking arion users who need systemd in their containers.

###### Notes

The test is rather small, so it makes sense to merge it into the `podman` test, but that file is changing with #123841

cc @NixOS/podman 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

roberth committed:
nixos/containers: Add virtualisation.containers.containersConf.cniPlugins

roberth committed:
nixos/podman: Add defaultNetwork.extraPlugins

roberth committed:
nixos/podman-dnsname: init

r-rmcgibbo commented:
Result of `nixpkgs-review pr 124589` at 0da67424 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>cniPlugins.dnsname</li>
    <li>nixos-install-tools</li>
  </ul>
</details>

zowoq commented:
I'm going to have several comments for this but I'm not going to have any time in the next few days, sorry.

roberth commented:
> I'm going to have several comments for this but I'm not going to have any time in the next few days, sorry.

I understand and I'm really grateful for your reviews on my earlier pr #123841. For me, having to switch from docker to podman was unexpected so I had to make time (and I hated that). I don't expect you to do the same.

roberth commented on nixos/tests/podman-dnsname.nix:
```diff
@@ -0,0 +1,42 @@
+import ./make-test-python.nix (
```
As mentioned, I'll happily merge this into the `podman` test after #123841 (unless you want it to be separate)

roberth reviewed (commented):
I've self-reviewed to highlight some of the decisio possible issues I've made. I hope that what remains will be trivial.

zowoq commented on nixos/modules/virtualisation/podman-dnsname.nix:
```diff
@@ -0,0 +1,46 @@
+in
+{
+  options = {
+    virtualisation.podman.dnsname = {
```
I'm not necessarily :-1:  on this but it doesn't seem quite right? By itself It's an option to add the `dnsname` plugin path to `containersConf` which we'll also have (manually) via `containersConf.cniPlugins`?

Could we do `virtualisation.podman.defaultNetwork.dnsname.enable` which adds the config and sets the plugin path?

I think that may also better align the options with your comment about a possible future rename?

roberth replied:
Done.

roberth commented on nixos/modules/virtualisation/podman.nix:
```diff
@@ -22,9 +23,23 @@ let
+      . + { "plugins": (.plugins + $extraPlugins) }
+    '';
+  } ''
+    jq <${pkgs.podman-unwrapped.src}/cni/87-podman-bridge.conflist \
```
The source derivation wasn't part of the output closure, but it's a good change nonetheless. With your pr, NixOS users won't have to download the podman sources for this one file.

roberth committed:
dnsname-cni: Use wrapper instead of patch for dnsmasq

The patch proved to be an incomplete solution while developing
nixosTests.podman-dnsname

roberth committed:
podman: Add nixosTests.podman-dnsname to tests

roberth commented:
It's ready for review again.

Turns out I can't read; `dnsname-cni` package already existed!

zowoq commented:
This looks good, thanks!
